read-fonts/src/lib.rs

```diff
@@ -71,6 +71,86 @@ pub(crate) mod codegen_prelude {
 
 include!("../generated/font.rs");
 
+#[derive(Clone)]
+/// Reference to the content of a font or font collection file.
+pub enum FileRef<'a> {
+    /// A single font.
+    Font(FontRef<'a>),
+    /// A collection of fonts.
+    Collection(CollectionRef<'a>),
+}
+
+impl<'a> FileRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        Ok(if let Ok(collection) = CollectionRef::new(data) {
+            Self::Collection(collection)
+        } else {
+            Self::Font(FontRef::new(data)?)
+        })
+    }
+
+    pub fn fonts(&self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a + Clone {
+        let copy = self.clone();
+        let len = match self {
+            Self::Collection(collection) => collection.len(),
+            _ => 1,
+        };
+        (0..len).filter_map(move |ix| match &copy {
+            Self::Font(font) => Some(Ok(font.clone())),
+            Self::Collection(collection) => collection.get(ix),
+        })
+    }
```

Comment:
Just for fun, my personal preferred way of writing this sort of thing is,

```rust
        let (iter_one, iter_two) = match self {
            FileRef::Font(font) => (Some(Ok(font.clone())), None),
            FileRef::Collection(collection) => (None, Some(collection.iter())),
        };
        iter_two.into_iter().flatten().chain(iter_one)
```

this is basically just at trick to ensure we have a single concrete return type.

Reply:
Interesting! I haven’t seen this one before